util/patch/patch.go

```diff
@@ -86,49 +83,167 @@ func (h *Helper) Patch(ctx context.Context, resource runtime.Object) error {
+//
+// This method has an internal backoff loop. When a conflict is detected, the method
+// asks the Client for the a new version of the resource we're trying to patch.
+//
+// Condition changes are then applied to the latest version of the object, and if there are
+// no unresolvable conflicts, the patch is sent again.
+func (h *Helper) patchStatusConditions(ctx context.Context, resource runtime.Object) error {
+	// Store the patch from the before/after object.
+	conditionPatch := conditions.NewPatch(
+		conditions.UnstructuredGetter(h.before),
+		conditions.UnstructuredGetter(h.after),
+	)
+
+	// Return early if there are no changes.
+	if conditionPatch.IsZero() {
+		return nil
+	}
 
-	if !reflect.DeepEqual(h.before, after) {
-		// only issue a Patch if the before and after resources (minus status) differ
-		if err := h.client.Patch(ctx, resource.DeepCopyObject(), h.resourcePatch); err != nil {
-			errs = append(errs, err)
-		}
+	// Make a copy of the object and store the key used if we have conflicts.
+	resourceCopy := resource.DeepCopyObject()
+	key, err := client.ObjectKeyFromObject(resourceCopy)
+	if err != nil {
+		return err
 	}
 
-	if (h.hasStatus || hasStatus) && !reflect.DeepEqual(h.beforeStatus, afterStatus) {
-		// only issue a Status Patch if the resource has a status and the beforeStatus
-		// and afterStatus copies differ
-		if err := h.client.Status().Patch(ctx, resource.DeepCopyObject(), h.statusPatch); err != nil {
-			errs = append(errs, err)
+	// Define and start a backoff loop to handle conflicts
+	// between controllers working on the same object.
+	//
+	// This has been copied from https://github.com/kubernetes/kubernetes/blob/release-1.16/pkg/controller/controller_utils.go#L86-L88.
+	backoff := wait.Backoff{
+		Steps:    5,
+		Duration: 100 * time.Millisecond,
+		Jitter:   1.0,
+	}
+	if err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		// Issue a patch.
+		err := h.client.Status().Patch(ctx, resourceCopy, h.conditionsPatch)
+		switch {
+		// Handle conflicts.
+		case apierrors.IsConflict(err):
+			// Only perform conflict resolution on objects that satisfy to the conditions.Setter interface.
+			conditionSetter, ok := resourceCopy.(conditions.Setter)
+			if !ok {
+				return false, errors.Wrapf(err, "cannot perform conflict resolution on unknown objects")
+			}
+
+			// Get a new copy of the object.
+			if err := h.client.Get(ctx, key, resourceCopy); err != nil {
+				return false, err
+			}
+
+			// Reset the condition patch before merging conditions.
+			h.conditionsPatch = client.MergeFromWithOptions(resourceCopy.DeepCopyObject(), client.MergeFromWithOptimisticLock{})
+
+			// Set the condition patch previously created on the new object.
+			if err := conditionPatch.Apply(conditionSetter); err != nil {
+				return false, err
+			}
+
+			// Requeue.
+			return false, nil
+		case err != nil:
+			return false, err
 		}
+
+		return true, nil
+	}); err != nil {
+		return err
+	}
+
+	return err
```

Comment:
```suggestion
	return nil
```

Or possibly remove the if around running `wait.ExponentialBackoff` and returning it's err value directly? Currently this appears to be returning `err` defined on line 167, which I would expect to be `nil` here.

Reply:
Went with the other suggestion, to return the error from the backoff loop directly